### DISCUSSION TITLE
nlpsol() code generation options

### DISCUSSION THREAD

**Discussion by waynekyrie** [casadi/casadi#3787] [Q&A] [answered]
Hi, is there any doc describing the options available for nlpsol().generation_dependencies()? I'm wondering if it is possible to reduce the size of the generated .c file by setting the options. Or are there other ways to reduce the size? Thanks!

**Comment by tmmsartor** [comment]
On the code generation side can try: 
```solver.generate_dependencies("nlp.c", {"verbose":False})```
but that will affect only file size, not "code" size or compilation time, if that is the real issue. (You can peak to more options [in the code](https://github.com/casadi/casadi/blob/2ab408eb582d7b62d00ed837f4d263ad885a0875/casadi/core/code_generator.cpp#L36) it seems they already have good defaults)

You can also try to use `expand` in your nlp definition, which expand MX to SX, that may also remove some packing/unpacking methods and may decrease code size. 
```solver = nlpsol("solver", "ipopt", nlp, {"expand":True})```

The ultimate way to go is to reduce the size of the computation graph of your nlp "dependencies" aka nlp cost, constraints and their derivatives. The generic advice is to keep your graphs compact by having a hierarchy of Functions, see also http://docs.casadi.org/#for-loop-equivalents.

As a side note to search for documentation my check list is: 
- Search on https://web.casadi.org/docs/
- Search on <casadi_repo>/docs/examples/ or <casadi_repo>/tests/ for proven usage
- Search on the forum here (and maybe also the old google group)
- Search on CasADi issues
- Search on CasADi wiki
- Search on the source code

**Reply by waynekyrie** [reply]
Thank you for the suggestions! I still have several questions,

1. By setting `"verbose":False`, which file size does it reduce? I do not see a difference between the generated c or cpp file sizes.
2. Does MX give a more compact graph than SX? By having a hierarchy, are you suggesting using MX in the main formulation and SX for each of the helper functions?

**Reply by waynekyrie** [reply]
I just tried modifying the problem formulation using hierarchy. The generated code size is now much smaller.

But I'm confused since I need to actually set `expand:False` to reduce the code size. Is it the correct behavior or I did something wrong?

**Reply by tmmsartor** [reply, marked as answer]
1) `"verbose":False` just avoid addition of comments, and it makes no difference when `"expand":True` is used because in that case comments are already removed.

2) If your bottle neck is graph size (construction time and compilation time)  using MX is a good idea, but just using MX doesn't necessarily make your graph more hierarchical  and compact, i.e. if one have 1x1 MX nodes with no hierarchy using MX would be likely worse. This [video section](https://youtu.be/Xfcp5RK7XZQ?t=1409) explain a bit those concepts.

3) Nothing wrong, it is totally expect to see smaller code with `"expand":False` if your expressions have some hierarchy, especially for large size. Having MX nodes brings some code overhead (like methods to perform packing and unpacking) so for "flat" graphs removing those overheads can have positive effect but for hierarchical matrix valued graphs the benefits of having fewer nodes should greatly outpace those overheads in terms of graph size and thus compilation time.

**Reply by waynekyrie** [reply]
Thank you! By doing hierarchy and setting `"expand`:False`, I'm now getting much smaller C files. 

But the compilation time is much longer now when compiling with optimization flags, `-O1` or `-O2` or `-O3`. The compilation time is fine when doing no optimization. Do you have any suggestions for reducing the compilation time as well?

**Reply by tmmsartor** [reply]
- It is expected that enabling optimization flag increase compilation time. 
- Decreasing compilation time of large source file may be not easy. 
- Just to understand compilation time was smaller before you restructured your code adding hierarchy and decreasing code size? 

- Adding more hierarchy i.e. by using foor-loop equivalent if you are not already doing this should decrease compilation time

- Another small thing you can try is to split functions compilation in different source file so that they can compiled in parallel, note that this will at best decrease compilation time by ~half, you can do that by adding something similar to the following code, after defining your `nlp` and `opts` dictionaries:
```python
w = nlp['x']
J = nlp['f']
g = nlp['g']

lam_f = SX.sym("lam_f")
lam_g = SX.sym("lam_g", g.sparsity())
param = SX.sym('param')

L = lam_f*J + dot(lam_g, g)
H, j = hessian(L,w)
Htriu = triu(H)

grad_f = Function("grad_f",[w, param],[J, gradient(J,w)])
jac_g = Function("jac_g",[w, param],[g, jacobian(g,w)])
hess_lag = Function("hess_lag", [w, param, lam_f, lam_g], [Htriu])

def compile_fun(fun):
  compiler = "gcc"
  flags = "-fPIC -shared -O3"
  src_file = fun.generate()
  so_file = src_file.replace(".c",".so")
  system(f"{compiler} {flags} {src_file} -o {so_file}")
  compiled_fun = external(fun.name(), so_file)
  return compiled_fun

# sequential
# [gf, jg, hl] = [compile_fun(fun) for fun in funs]
# parallel
pool = Pool(processes=3)
funs = [grad_f, jac_g, hess_lag]
[gf, jg, hl] = pool.map(compile_fun, funs)

opts.update({'grad_f':gf, 'jac_g':jg, 'hess_lag':hl})

solver = nlpsol('solver', 'ipopt', nlp, opts)
```

**Reply by waynekyrie** [reply]
I have two versions, with and without hierarchy. The generated code file for the one with hierarchy is about 10Mb and the other one without hierarchy is about 60Mb. So yes the file size is smaller. And when compiling them with `-O0`, the smaller one is faster. But when I compile with an optimization flag, then the hierarchy one with smaller size seems taking longer time. It actually never finishes after waiting for hours so I do not know the exact time difference.

I will try the for-loop equivalent. Thank you for your insights!

**Reply by tmmsartor** [reply]
Ok I see I would guess than the version without hierarchy would take even longer to compile.
Yes if the generated source is 10Mb won't be surprising if it takes an  inconveniently long time to compile.
Other options for you is to use for-loop equivalent (see also [this issue](https://github.com/casadi/casadi/issues/3490) for examples) and optimize more your computation graph.

Also try to add `{"cse":True}` in your `Function` options, which will eliminate common sub-expressions (as in sub-graphs which repeat themself), this can also make your graph smaller.

**Reply by waynekyrie** [reply]
Will try. Thank you so much for your suggestions!
